src/service/position-observer/position-observer-impl.js

```diff
@@ -0,0 +1,303 @@
+    this.inScroll_ = false;
+
+    /** @private {boolean} */
+    this.measure_ = false;
+
+    /** @private {function()} */
+    this.boundStopScroll_ = debounce(this.win_, () => {
+      this.inScroll_ = false;
+    }, 500);
+
+    /** @private {boolean} */
+    this.needRefreshOnMessage_ = false;
+  }
+
+  /**
+   * @param {!Element} element
+   * @param {PositionObserverFidelity} fidelity
+   * @param {function(?./position-observer-entry.PositionInViewportEntryDef)} handler
+   */
+  observe(element, fidelity, handler) {
+    const entry = new PositionObserverEntry(element, fidelity, handler);
+
+    this.entries_.push(entry);
+
+    if (this.entries_.length == 1) {
+      this.startCallback_();
+    }
+
+    this.vsync_.measure(() => {
+      this.updateSingleEntry_(entry);
+    });
+  }
+
+  /**
+   * @param {!Element} element
+   */
+  unobserve(element) {
+    for (let i = 0; i < this.entries_.length; i++) {
+      if (this.entries_[i].element == element) {
+        this.entries_[i].handler = null;
+        this.entries_.splice(i, 1);
+        if (this.entries_.length == 0) {
+          this.stopCallback_();
+        }
+        return;
+      }
+    }
+    dev().error(TAG, 'cannot unobserve unobserved element');
+  }
+
+  /**
+   * Callback function that gets called when start to observe the first element.
+   * @private
+   */
+  startCallback_() {
+    this.viewport_.connect();
+    // listen to viewport scroll event to help pass determine if need to
+    this.unlisteners_.push(this.viewport_.onScroll(() => {
+      this.onScrollHandler_();
+    }));
+    this.unlisteners_.push(this.viewport_.onResize(() => {
+      this.onResizeHandler_();
+    }));
+    this.unlisteners_.push(this.viewport_.onHostMessage(event => {
+      this.onHostMessageHandler_(event);
+    }));
+  }
+
+  /**
+   * Callback function that gets called when unobserve last observed element.
+   * @private
+   */
+  stopCallback_() {
+    this.viewport_.disconnect();
+    while (this.unlisteners_.length) {
+      const unlisten = this.unlisteners_.pop();
+      unlisten();
+    }
+  }
+
+  /**
+   * @param {!Element} element
+   * @param {PositionObserverFidelity} fidelity
+   */
+  changeFidelity(element, fidelity) {
+    for (let i = 0; i < this.entries_.length; i++) {
+      const entry = this.entries_[i];
+      if (entry.element == element) {
+        entry.fidelity = fidelity;
+        if (fidelity == PositionObserverFidelity.HIGH) {
+          entry.turn = 0;
+        }
+        return;
+      }
+    }
+    dev().error(TAG, 'cannot change fidelity on unobserved element');
+  }
+
+  /**
+   * This should always be called in vsync.
+   * @param {boolean=} force
+   * @param {boolean=} opt_remeasure
+   * @visibleForTesting
+  */
+  updateAllEntries(force, opt_remeasure) {
+    for (let i = 0; i < this.entries_.length; i++) {
+      const entry = this.entries_[i];
+      if (opt_remeasure) {
+        entry.element.inIframePositionRect = null;
```

Comment:
😢  I don't like this either. But in the case of inabox, I know an element's relative pos to the inabox window is rarely changed. So I store the value in `elemenet['inIframePositionRect']` (need avoid using .) to avoid unnecessary `getBoundingClientRect()` call. However in the case of inabox window resize I'll need to clear the stored value. That's why I set it back to null here.